lighthouse-core/gather/driver/wait-for-condition.js

```diff
@@ -0,0 +1,420 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const log = require('lighthouse-logger');
+const LHError = require('../../lib/lh-error.js');
+const ExecutionContext = require('./execution-context.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/** @typedef {import('../../lib/network-recorder.js')} NetworkRecorder */
+/** @typedef {{promise: Promise<void>, cancel: function(): void}} CancellableWait */
+
+/**
+ * @typedef WaitOptions
+ * @prop {number} pauseAfterFcpMs
+ * @prop {number} pauseAfterLoadMs
+ * @prop {number} networkQuietThresholdMs
+ * @prop {number} cpuQuietThresholdMs
+ * @prop {number} maxWaitForLoadedMs
+ * @prop {number|undefined} maxWaitForFcpMs
+ * @prop {{waitForFcp: typeof waitForFcp, waitForLoadEvent: typeof waitForLoadEvent, waitForNetworkIdle: typeof waitForNetworkIdle, waitForCPUIdle: typeof waitForCPUIdle}} [_waitForTestOverrides]
+ */
+
+/**
+ * Returns a promise that resolves immediately.
+ * Used for placeholder conditions that we don't want to start waiting for just yet, but still want
+ * to satisfy the same interface.
+ * @return {{promise: Promise<void>, cancel: function(): void}}
+ */
+function waitForNothing() {
+  return {promise: Promise.resolve(), cancel() {}};
+}
+
+/**
+ * Returns a promise that resolve when a frame has been navigated.
+ * Used for detecting that our about:blank reset has been completed.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @return {CancellableWait}
```

Comment:
this is the only return type change, being more consistent felt like it belonged here even if we don't use `cancel` yet